### ISSUE TITLE
[Question/Bug] `grad_accum_steps` is used by the effective batch-size check but not applied in Wan21 camera trainers

### ISSUE THREAD

**Issue by Ohmuskmelon** [shengshu-ai/minWM#12]
Hi, thank you for open-sourcing minWM.

While preparing a training smoke test for the Wan2.1 Action2V pipeline, I noticed a possible mismatch between the effective batch-size check and the actual optimizer-step behavior.

## Repository version

```text
commit: df522a26cd4409d3e3e8f269cc98eac069b5df47
pipeline: Wan21 / Action2V
```

## Effective batch-size check

At this commit, `Wan21/wan_train.py` contains:
    
    
    total_gpus = int(os.environ.get("WORLD_SIZE", 1))
    dp = total_gpus // config.sp_size
    ga = getattr(config, "grad_accum_steps", 1)
    bsz = dp * ga

The assertion below this code suggests increasing `grad_accum_steps` when the effective batch size is below 16.

Two details stood out to me:

1. `grad_accum_steps` appears to be used by this guard, but I could not find corresponding gradient accumulation in the target camera trainers.
2. The general effective-batch formula would normally also include the per-device batch size:
    
    
    effective_batch_size = dp * config.batch_size * grad_accum_steps

The official camera configs I checked currently use `batch_size: 1`, so the missing factor does not change the numerical result for those configs, but it may matter if `batch_size` is changed.

## Optimizer-step behavior

For example, in `Wan21/wan_trainer/camera_bidirectional_diffusion.py`, each call to `train_one_step()` performs:
    
    
    self.generator_optimizer.zero_grad()
    generator_loss.backward()
    self.generator_optimizer.step()
    self.step += 1

The corresponding training loop consumes one dataloader batch before each call.

I observed the same one-batch-per-update structure in the camera AR trainer. The camera consistency-distillation and DMD trainers inherit outer loops that also zero gradients, execute one forward/backward call, and step the relevant optimizer without consulting `grad_accum_steps`.

I could not find an accumulation loop, loss scaling by `grad_accum_steps`, or optimizer-step boundary logic conditioned on `grad_accum_steps`.

I also did not find `no_sync()` or an equivalent optimization for non-boundary microsteps, although this would affect distributed communication efficiency rather than whether accumulation is functionally performed.

## Example

With:
    
    
    WORLD_SIZE=8
    sp_size=1
    batch_size=1
    grad_accum_steps=2

the current check reports:
    
    
    dp = 8
    effective_bsz = 16

and the assertion passes.

However, unless two distinct dataloader microbatches are accumulated before `optimizer.step()`, the optimizer update appears to use:
    
    
    8 data-parallel ranks × 1 sample per rank = 8 samples

rather than 16 samples.

## Config naming

I also noticed inconsistent configuration keys:
    
    
    Wan21/configs/bidirectional_camera.yaml:
      grad_accum_steps: 1
    
    Wan21/configs/ar_camera_tf.yaml:
      grad_accum_steps: 1
    
    Wan21/configs/causal_cd_camera.yaml:
      grad_accum: 1

`wan_train.py` reads `grad_accum_steps`, and I could not find a target trainer that consumes `grad_accum`. Therefore, `grad_accum` in `causal_cd_camera.yaml` appears not to affect either the guard or the optimizer loop.

`causal_forcing_dmd_camera.yaml` appears to define neither key and therefore uses the guard's default value of 1.

## Questions

Could you please clarify the intended official behavior?

1. Is gradient accumulation implemented elsewhere that I may have missed?
2. Is `grad_accum` in `causal_cd_camera.yaml` intended to be `grad_accum_steps`?
3. Is the official setup expected to achieve a batch of at least 16 through data-parallel batch size alone?
4. For an 8-GPU setup with `batch_size=1`, should `grad_accum_steps=2` consume two forward/backward microbatches before one optimizer update?
5. Should gradient clipping, EMA updates where applicable, logging/global steps, checkpoint intervals, and `self.step` operate at the optimizer-step boundary?

I did not find an optimizer LR scheduler in these target trainers. If one is added or used elsewhere, I would expect it to advance at the optimizer-step boundary as well.

## Expected behavior, if accumulation is intended

For a simple trainer and `grad_accum_steps=2`, I would expect behavior approximately equivalent to:
    
    
    optimizer.zero_grad()
    
    for microstep in range(2):
        batch = next(dataloader)
        loss = compute_loss(batch) / 2
        loss.backward()
    
    clip_grad_norm(...)
    optimizer.step()
    self.step += 1

EMA updates, optimizer-dependent logging, checkpoint intervals, and global-step advancement would then occur once at this boundary.

For DMD, what is the intended interaction between gradient accumulation and `dfake_gen_update_ratio`? Should the generator and critic/fake-score optimizers each maintain their own accumulation window?

## Local verification

After implementing this behavior locally, I completed a reduced-option, one-optimizer-step smoke test for the camera bidirectional stage on 8 GPUs:

- two distinct real dataloader microsteps per rank;
- finite losses;
- finite, nonzero gradients;
- one optimizer step;
- a nonzero update to a representative trainable parameter;
- one `self.step` increment.

This was only a one-step training-semantics smoke test, not a convergence or full-training reproduction. I would like to confirm the intended official semantics before relying on the modification or preparing a pull request.

Thank you.

**Comment by zhuhz22** [comment]
Yes that's exactly a potential bug, and we'll fix it later. Thx!

**Comment by zhuhz22** [comment]
Solved in this new version！

**Comment by Ohmuskmelon** [comment]
Just tested on the new version — confirmed fixed. Thanks for the quick turnaround!
